### PR TITLE
Add support for | to StrawberryUnion

### PR DESCRIPTION
**Pull request by patrick91**
This PR adds support for the new union operator to `StrawberryUnion` :)

Supports:
1. Merging with `None` (making a union optional) -> `MyUnion | None`
2. Merging with another union -> `MyUnion | AnotherUnion`
3. Merging with another type -> `MyUnion | AnotherType`
4. Merging with a union -> `MyUnion | Union[AnotherType, YetAnotherType]`

### CODE REVIEW

File: strawberry/union.py [hunk 1]
```diff
@@ -63,6 +63,32 @@ def __hash__(self) -> int:
...
+
+        if isinstance(other, StrawberryUnion):
+            return StrawberryUnion(
+                type_annotations=tuple(
+                    set(
+                        (
+                            *self.type_annotations,
+                            *other.type_annotations,
+                        )
+                    )
```
Comment (patrick91) [inline strawberry/union.py:83]: shall we make type_annotations always a set? 🤔

**Comment by patrick91** [comment]
@BryceBeagle waht do you think of this? :D

File: strawberry/union.py [hunk 2]
```diff
@@ -63,6 +64,35 @@ def __hash__(self) -> int:
...
         return id(self)
 
+    def __or__(self, other: Union[StrawberryType, type]) -> StrawberryType:
+        if hasattr(other, "_type_definition"):
+            return StrawberryUnion(
+                type_annotations=(
+                    *self.type_annotations,
+                    StrawberryAnnotation(other),
+                )
+            )
```
Comment (BryceBeagle) [inline strawberry/union.py:74]: Maybe we can make a factory method for this? `StrawberryUnion.extend` or something?

File: strawberry/union.py (hunk 1 above)
Comment (BryceBeagle) [reply strawberry/union.py:87]: According to [PEP 604](https://www.python.org/dev/peps/pep-0604/):
> The order of the items in the Union should not matter for equality.
> ```
> (int | str) == (str | int)
> (int | str | float) == typing.Union[str, float, int]
> ```
so I think that's a good idea. Only issue could be hashability of the annotations

File: tests/types/resolving/test_union_pipe.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    UserOrError = strawberry.union("UserOrError", (User, Error))
+    ExampleOrError = strawberry.union("ExampleOrError", (Example, Error))
+    annotation = StrawberryAnnotation(UserOrError | ExampleOrError)
+    resolved = annotation.resolve()
+
+    assert isinstance(resolved, StrawberryUnion)
+
+    assert len(resolved.type_annotations) == 3
+
+    # doing this to avoid errors due to order of types
```
Comment (BryceBeagle) [inline tests/types/resolving/test_union_pipe.py:131]: This shouldn't be an issue if we make the annotations a set, right?

File: tests/types/resolving/test_union_pipe.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
+import sys
```
Comment (BryceBeagle) [inline tests/types/resolving/test_union_pipe.py:1]: Rename the file to `test_unions_pipe.py` (or maybe `test_union_pipes.py`) for plurality consistency with the other files

**Review by BryceBeagle — CHANGES_REQUESTED** [review]

**Comment by patrick91** [comment]
I'm going to close this since we'll support unions with Annotated soon!

**Pull request closed without merging**